### ISSUE TITLE
Document HTTP proxying

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7381]
|<img alt="etrepum's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| etrepum reported|
|-|-|
|Trac ID|trac#399|
|Type|enhancement|
|Created|2003-11-17 03:40:34Z|
|Branch|https://github.com/twisted/twisted/tree/proxy-docs-399|

<details><summary>Searchable metadata</summary>

```
trac-id__399 399
type__enhancement enhancement
reporter__etrepum etrepum
priority__normal normal
milestone__ 
branch__branches_proxy_docs_399 branches/proxy-docs-399
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__web web
keywords__documentation documentation
time__1069040434000000 1069040434000000
changetime__1469343154301186 1469343154301186
version__ 
owner__jesstess jesstess
cc__itamarst cc__etrepum cc__doko cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="etrepum's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|etrepum commented|
|-|-|

```
#!html
<pre>
I don't see any documentation or tests for using a HTTP proxy, only creating a HTTP proxy server.  I 
don't see any tests for the HTTP 
proxying at all really.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I had a issue open for this somewhere (writing tests).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Issue238 is the writing tests one, making you both nosy on that.

Retitling this bug to be the documentation bug and assigning to me.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [33249]) Branching to 'proxy-docs-399'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [33253]) Add more proxy examples to the web examples page.

refs [#7381](https://github.com/twisted/twisted/issues/7381).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess commented|
|-|-|

(In [33254]) Add a section on proxies and reverse proxies to the using-twistedweb howto.

refs [#7381](https://github.com/twisted/twisted/issues/7381).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess removed owner|
|-|-|

This branch:

 * Adds 2 examples of forward proxies to doc/web/examples/
 * Adds a section on proxies and reverse proxies to doc/web/howto/using-twistedweb

I wanted to talk about ```ReverseProxy``` as a parallel to ```Proxy```, but I think that code is broken. I've filed [#5417](https://github.com/twisted/twisted/issues/5417) for that and mostly dodge the issue in favor of talking about ```ReverseProxyResource``` in the howto. We can open a ticket about adding actual ```ReverseProxy``` examples once that code works and has tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks!  Documentation is awesome.

  1. The paragraph beginning _In this example we use_ goes on to say _to server a ReverseProxyResource_ - typo on what should be _serve_.
  1. Might be good to be consistent about logging amongst the three proxying examples.  It shouldn't need to be enabled in `logging-proxy.py` for the printed output to show up.  The extra log noise in that one is a bit annoying too.
  1. doc/web/examples/proxy.py has a line with some trailing whitespace that should be removed
  1. doc/web/examples/index.html links to `proxy.py` twice and not to `logging-proxy.py` at all
  1. it also incorrectly describes `reverse-proxy.py` as based on `ReverseProxy` rather than `ReverseProxyResource`.
  1. Maybe the _ReverseProxy_ heading in `using-twistedweb.xhtml` should be _ReverseProxyResource_ as well.

Otherwise looks great.  The [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/proxy-docs-399&num_builds=10) look good so far.  Please merge when the above points are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess commented|
|-|-|

(In [33261]) Address review comments by exarkun.

refs [#7381](https://github.com/twisted/twisted/issues/7381).

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess set status to `closed`|
|-|-|

(In [33262]) Merge proxy-docs-399

Author: jesstess
Reviewer: exarkun
Fixes: [#7381](https://github.com/twisted/twisted/issues/7381)

Add a web howto section on proxies and reverse proxies and add more
proxy examples to the web examples directory.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|GitHub commented|
|-|-|

In changeset 2cdecce902af9b808af7195dd1c42d662acb7210
```
#!CommitTicketReference repository="" revision="2cdecce902af9b808af7195dd1c42d662acb7210"
Merge pull request #399 from twisted/8661-einarfd-ssh-session

Author: einarfd
Reviewer: rodrigc
Fixes: #8661

Port twisted.conch.ssh.session to Python 3
```
